File: store/279.完全平方数.c
/*
 * @lc app=leetcode.cn id=279 lang=c
 *
 * [279] 完全平方数
 */

// @lc code=start

int numSquares(int n)
{
  int i, j, *dp;
  dp = (int *)malloc(sizeof(int) * (n + 1));
  for (i = 0; i < n + 1; i++)
    dp[i] = INT_MAX;

  dp[0] = 0;

  for (i = 1; i * i < n + 1; i++)
    for (j = i * i; j < n + 1; j++)
      dp[j] = dp[j] < dp[j - i * i] + 1 ? dp[j] : dp[j - i * i] + 1;

  return dp[n];
}
// @lc code=end

File: store/279.完全平方数.js
/*
 * @lc app=leetcode.cn id=279 lang=javascript
 *
 * [279] 完全平方数
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function (n) {
  let dp = new Array(n + 1).fill(Number.MAX_SAFE_INTEGER);
  dp[0] = 0;

  for (let i = 1; i * i <= n; i++) {
    let num = i * i;
    for (let j = num; j <= n; j++) {
      dp[j] = Math.min(dp[j], dp[j - num] + 1);
    }
  }

  return dp[n];
};
// @lc code=end
